### ISSUE TITLE
Should search be case-sensitive?

### ISSUE THREAD

**Issue by psads-git** [mike-fabian/ibus-typing-booster#251]
While I am in process of changing the timestamps of my database, I have something that I would like to suggest to you: To make search case-_insensitive_. For instance, if I write mike, nothing is found by ibus-typing-booster, but if I start writing Mi... the Mike suggestion emerges immediately. I think it would be useful to be possible to write mi... and then the suggestion Mike being immediately offered. 

What do you think about this?

**Comment by mike-fabian** [comment]
Yes, I think search should be case-insensitive and if this is possible it should be case insensitive by default.

**Comment by mike-fabian** [comment]
I don’t want to make this optional though, I think it should **always** be case insensitive.

**Comment by psads-git** [comment]
Thanks, Mike. That will significantly improve ibus-typing-booster.

**Comment by mike-fabian** [comment]
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/ has 2.14.18 builds now which have the case insensitive matching.

**Comment by psads-git** [comment]
Thanks, Mike. This version of ibus-typing-booster is almost unusable, since the search is extremely slow. 

Let me add that my database has about 70MB (I input into it several books, to help the prediction).

**Comment by mike-fabian** [comment]
Is it slower than before?

**Comment by psads-git** [comment]
Yes, Mike, several orders of magnitude slower.

**Comment by mike-fabian** [comment]
But probably because of your huge database, or do you think the case insensitive matching change anything?

**Comment by mike-fabian** [comment]
If you downgrade to 2.14.17, is that faster?

**Comment by psads-git** [comment]
If I downgrade to  2.14.17, ibus-typing-booster is very fast. 

I do not have any solid idea why is that happening, but, since the result of each search has many more elements, the ordering by use frequency may be extremely slow, because of the number of elements that it has to order -- if I remember well, the sorting algorithms are  O(n^2) (or something close to that).

**Comment by mike-fabian** [comment]
> Let me add that my database has about 70MB (I input into it several books, to help the prediction).

By the way, I found reading books into Typing Booster not to be helpful. For example, in 2014, I let Typing Booster read the “Hitchhikers Guide to the Galaxy” and “The Picture of Dorian Gray”. It didn’t seem to help my typing at all. I could easily open any page in one of these books and type any sentence from these books easily with good prediction, but what I really wanted to type (almost) never seemed to occur in these books. I guess this is true for almost any book except possibly if you wrote it yourself. User input is so variable, it doesn’t help much to read text from other writers, even a huge amount of such text doesn’t seem to help. 

Now when I implemented the expiry of old entries and looked at what entries were expired, I noticed that a most entries from both these books had been created in 2014 and have **never** been touched since then. 

So I think reading books doesn’t help much. 

Maybe I should rethink how to read huge texts, I have no good idea at the moment though. Learning from your own input is helpful, learning from other peoples input seems to have very limited value. Maybe, when reading a text above a certain size, I should only add stuff from that text to the database if it really appears very often. That still might not help very much, for example “The picture of Dorian Gray” contains 47 times the text “said Lord Henry”. I will probably never type that, I will probably type “said” often but “Lord” and “Henry” very rarely and it is unlikely that I ever type the complete text “said Lord Henry”.

**Comment by mike-fabian** [comment]
Now I have measured and indeed it is much slower.

My database has 140000 rows and a size of 10MB and with that it is about 10 times slower. 

I measured both the time it takes to do the case insensitive lookup in the hunspell dictionaries and in the database.

The difference for the lookup in the dictionaries is insignificant, the difference in speed between the case insensitive regular expressions used now and the `.startswith()` used before seems to be within measurement error.

But the difference when doing the database lookup is huge. As soon as I use

```
PRAGMA case_sensitive_like = false;
```

for the sqlite database, it becomes about 10 times slower for me.

So probably there are two things slowing this down:

- the case insensitive LIKE operator is probably slow already

- When doing the LIKE case insensitive, the number of records matched is probably a lot bigger and then calculating the linear combinations of the counts of the previous two words has a lot more work to do and becomes much slower

**Comment by psads-git** [comment]
So, Mike, case-insensitiveness is perhaps a bad idea! ;-)

**Comment by mike-fabian** [comment]
Some examples from my current database:

```
sqlite> PRAGMA case_sensitive_like = true;
sqlite> select sum(user_freq) from phrases where input_phrase like "Thi%" ;
411
sqlite> PRAGMA case_sensitive_like = false;
sqlite> select sum(user_freq) from phrases where input_phrase like "Thi%" ;
2768
sqlite>

sqlite> PRAGMA case_sensitive_like = true;
sqlite> select sum(user_freq) from phrases where input_phrase like "Th%" ;
2712
sqlite> PRAGMA case_sensitive_like = false;
sqlite> select sum(user_freq) from phrases where input_phrase like "Th%" ;
18887
sqlite> 

sqlite> PRAGMA case_sensitive_like = true;
sqlite> select sum(user_freq) from phrases where input_phrase like "Te%" ;
188
sqlite> PRAGMA case_sensitive_like = false;
sqlite> select sum(user_freq) from phrases where input_phrase like "Te%" ;
2199
sqlite> 

sqlite> PRAGMA case_sensitive_like = true;
sqlite> select sum(user_freq) from phrases where input_phrase like "te%" ;
1998
sqlite> PRAGMA case_sensitive_like = false;
sqlite> select sum(user_freq) from phrases where input_phrase like "te%" ;
2199
sqlite>
```

Looks like the amount of entries returned by a case insensitive LIKE operator is really a lot bigger.

Probably we cannot do case insensitive matching in the database then.

Maybe only in the dictionaries.

So for example when using the en_US dictionary, “mike” and “Mike” would both match:

```
$ grep -i mike /usr/share/myspell/en_US.dic 
Mike/M
mike/MGDS
mfabian@taka:~
$
```

but not when  using the en_GB dictionary:

```
mfabian@taka:~
$ grep -i mike /usr/share/myspell/en_GB.dic 
mike/DMGS
mfabian@taka:~
$ 
```

It looks like I can do the case insensitive match in the dictionaries without measurable slowdown and at least that makes it possible to type “corme” or “Corme” and get “Cormeilles-en-Parisis” when using the French dictionary:

```
mfabian@taka:~
$ grep Paris /usr/share/myspell/fr_FR.dic 
Cormeilles-en-Parisis
Paris
Seyssinet-Pariset
Tout-Paris
mfabian@taka:~
$ 
```

**Comment by mike-fabian** [comment]
Would a “halfway case insensitive” solution, i.e. case insensitive in the dictionaries but not in the database as described above be good?

**Comment by psads-git** [comment]
Yes, Mike, a  “halfway case insensitive” solution would definitely be better than nothing. And to use case insensitiveness in the database is, as is very clear, impractical.

**Comment by mike-fabian** [comment]
I just noticed that case insensitive matching in the database is even possible with no loss in performance at alŀ if I do it the same way I do it for accent  insensitive matching. 

Accent insensitive matching in the database is currently done like this:

Before the input the user has typed (`input_phrase`) is saved to the database all accents from the user input are removed, i.e. the `input_phrase` is always stored without accents in the database. 

If the user then  types the something again, the accents are removed from the input and then the match against the database is done. 

This works well and does not cause any loss in performance. The disadvantage is of course that this is an option one cannot switch immediately. When I make accent sensitive matching in the database configurable as discussed in 

https://github.com/mike-fabian/ibus-typing-booster/issues/231 

Then a change in that option will only be effective for new input after that option was changed. Old rows in the database cannot be changed anymore, they would stay as they were. 

One could do the same with upper and lower case: Always store `input_phrase` in lower case in the database, always convert user input to lower case before matching. 

That would be fast.

But if this would be added as an option in the setup tool, then changing that option would only have an effect on new input.

**Comment by psads-git** [comment]
Thanks, Mike, but I am not sure whether it will work. Suppose that one wants to write 

`Mike` 

and types `mike`. How can ibus-typing-booster automatically transform `mike` into `Mike`?

My idea of case-insensitive search was to avoid typing uppercase letters mainly in names.

I believe that you are thinking in something like the following: One types

`Mi` (notice the capital `M`)

and ibus-typing-booster will suggest 

`Mike`

This would be a great progress though, as it would allow a substantial size-reduction of the database while maintaining the same prediction performance.

**Comment by mike-fabian** [comment]
> Thanks, Mike, but I am not sure whether it will work. Suppose that one wants to write
> 
> `Mike`
> 
> and types `mike`. How can ibus-typing-booster automatically transform `mike` into `Mike`?

That can be done easily in Python:

```
mfabian@taka:~
$ python3
Python 3.10.0 (default, Oct  4 2021, 00:00:00) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 'Mike'.lower()
'mike'
>>> 'Mike'.upper()
'MIKE'
>>> 'mike'.title()
'Mike'
>>> 'mike'.upper()
'MIKE'
>>> 'mIKe'.upper()
'MIKE'
>>> 'mIKe'.lower()
'mike'
>>> 'mIKe'.title()
'Mike'
>>> 
```

> My idea of case-insensitive search was to avoid typing uppercase letters mainly in names.
> 
> I believe that you are thinking in something like the following: One types
> 
> `Mi` (notice the capital `M`)
> 
> and ibus-typing-booster will suggest
> 
> `Mike`

I think that you can type either `Mi` or `mi` and in both cases you will get the suggestion `Mike`.

That is almost the same as with the accents, for example my database currently contains:

```
sqlite> select * from phrases where input_phrase  == "deja";
426586|deja|déjà|tu|As|1|1636959600.50043
426719|deja|déjà assisté|tu|as|1|1636960347.05661
427645|deja|déjà|||2|1636983062.65822
sqlite> 
```

The input_phrase column always contains `deja` without the accents, even when I actually typed it with the accents.

When saving the the database, that column is always saved without the accents (There are some language dependent exceptions, we talked about these before, but let’s ignore these for this explanation). And from the user input the accents are also removed before attempting a match in the database. That means no matter whether the user types `de` or `dé`, both will match the above rows.

One could do the same with upper and lower case. For example when the user currently types `Mi` and selects `Mike`, or types `mi` and selects `Mike`, that would create two database entries like this:

```
426586|Mi|Mike|||1|1636959600.50043
426719|mi|Mike|||1|1636960347.05661
```

If the user input was always converted to lower case when when saving to the database and when matching against the database, one would have only one entry with count 2 like this:

```
426719|mi|Mike|||2|1636960347.05661
```

and typing `Mi` would still match that row because `Mi` would be converted to `mi` before trying the match.

But switching that option would only have an effect of new entries. The old ones would very slowly disappear because of the expiry.

Theoretically one could convert the database to all lower case input_phrase automatically when the option `case insensitive matching` is switched on.

But when it is switched off again, there would be no way to automatically convert it back as some information was lost when doing the conversion to lower case.

I think it might be better not to attempt any conversion of the existing rows at all and let the value of that option only have an effect on new rows.

> This would be a great progress though, as it would allow a substantial size-reduction of the database while maintaining the same prediction performance.

I am not sure whether this is a good idea or not, probably I would make it an option if I implement this. I think the reduction in the number of rows would not be substantial, there are nto so many words which usually start with upper case in most languages. There are more in German because all nouns start with upper-case in German, but even in the case of German I think the number or rows which would be merged into one because of this would be rather small.

Whether prediction performance gets better or worse is partly a matter of opinion. If you type more exact, the prediction can be more exact. If you always type upper-case letters correctly, then the prediction has less choice. It is the same as with the accents, I think. If you type `de` now, words starting with `de` and `dé` can be predicted because we use accent insensitive matching. Accent sensitive matching would reduce the number of matching candidates.

This is exactly the same with case insensitive matching, it increases the number of matching candidates.

If there are more candidates, it may take more time to scroll through the candidate list and select the correct one.

So in the long run, both accent insensitive matching and case insensitive matching should be options. Both on by default probably.

**Comment by psads-git** [comment]
Thanks, Mike. That is not only names that need to be capitalized, but all first words after a period! Example:

`This is a period. Now is the first word after period.`

So, the need of capitalization is very recurrent!

**Comment by mike-fabian** [comment]
There is an autocapitalization feature after sentence endings.

**Comment by psads-git** [comment]
Thanks, Mike, for remembering me that feature, which I had, meanwhile, forgotten about.

**Comment by mike-fabian** [comment]
By the way, with the current database limited to 50000 rows, doing case insensitive matching by using 

```
 PRAGMA case_sensitive_like = false;
```

causes a slowdown of about a factor of 2. Enough that I can notice the slowdown while typing.

The advantage of this method is of course that one could switch the option in an instant and it would have an effect on all rows existing in the database.

Whereas the other method of storing input phrases only in lower case in the database and converting each new input to lower case before doing the matching would effect only new rows. 

1. Use PRAGMA `case_sensitive_like` = false;
 - Disadvantage: Slowdown by a factor of 2
 - Advantage: If this is made into an option, it can be switched immediately
2. Use python lower()
 - Disadvantage: If this is made into an option, switching the option effects only new rows
 - Advantage: No slowdown

I tend to think 2. is better because one will probably not switch that option very often. One will probably figure out what setting one likes and then just keep that. So one will  have the disadvantage that it  effects only new rows for a limited time but will have the speed advantage forever.

And I am not even sure whether I want to make this into an option, maybe just do case insensitive matching always with no option to switch to case sensitive matching. I don’t know yet whether I should add this as an option or not.

First I’ll make a build with case insensitive matching now with no option to switch if off and let you test and  hear your opinion.

**Comment by mike-fabian** [comment]
While doing the case insensitive match according to method 2. I found this small problem when learning by reading from text files which contain accented words:

https://github.com/mike-fabian/ibus-typing-booster/issues/252

**Comment by psads-git** [comment]
Thanks, Mike. I also prefer the option 2 (python lower()). A power user can change all database entries to lowercase...

And I do not see any important reason why to leave optional (to the user) the search to be case-insensitive or case-sensitive.

**Comment by psads-git** [comment]
> While doing the case insensitive match according to method 2. I found this small problem when learning by reading from text files which contain accented words:
> 
> #252

That is not a big problem, Mike!

**Comment by mike-fabian** [comment]
> And I do not see any important reason why to leave optional (to the user) the search to be case-insensitive or case-sensitive.

I think I also prefer not to add an option to switch this off. Let’s see what we think after testing it for a while.

I already implemented option 2. and found that it does not cause any slowdown (Actually it was even slightly faster in my measurement but that was well within measurement error, I think it certainly is a tiny bit slower, but not enough to be easily measurable)

**Comment by mike-fabian** [comment]
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/ has 2.14.20 builds  now which have case insensitive matching both in the dictionaries and in the database and they also have a fix for 

https://github.com/mike-fabian/ibus-typing-booster/issues/252

**Comment by psads-git** [comment]
Thanks, Mike. I will try and will let you know whether something goes wrong.

This may be not related, but, the option `auto capitalize` does not work when typing one's answer here in GitHub.

**Comment by mike-fabian** [comment]
> This may be not related, but, the option `auto capitalize` does not work when typing one's answer here in GitHub.

It works even here after a sentence end character. I.e. if you type `test . test ` you will get `test. Test `. 

It does not work if you move the focus with the mouse to some other window and come back to the github comment and place the cursor after a sentence end character. Neither does it work if you move the cursor with the arrow keys.   And it doesn't work if you place the cursor at the beginning of the github comment field. All these use cases when changing the cursor position would need well working surrounding text.  Typing Booster must be able to read the surrounding text and find out whether there is a sentence end character or the beginning of the entry box left of the cursor. This does not work well with the current broken surrounding text support. 

Currently, `self._new_sentence` is set to True only when a non-empty commit ends with a sentence end character:

```
        if not commit_phrase.isspace():
            # If the commit space contains only white space
            # leave self._new_sentence as it is!
            self._new_sentence = False
            if itb_util.text_ends_a_sentence(commit_phrase):
                self._new_sentence = True
```

And it is reset to False again in do_reset():

```
    def do_reset(self) -> None:
        '''Called when the mouse pointer is used to move to cursor to a
        different position in the current window.

        Also called when certain keys are pressed:

            Return, KP_Enter, ISO_Enter, Up, Down, (and others?)

        Even some key sequences like space + Left and space + Right
        seem to call this.

        '''
        ...
        self.clear_context()
        ...
```

and in do_focus_in():

```
    def do_focus_in(self) -> None:
        '''Called when a window gets focus while this input engine is enabled

        '''
        ...
        self.clear_context()
        ...
```

both do_reset() and do_focus_in() call clear_context() which resets the self._new_sentence:


```
    def clear_context(self) -> None:
        '''Clears the context stack which remembers the last two words typed
        '''
        if DEBUG_LEVEL > 1:
            LOGGER.debug(
                'context=“%s” “%s” “%s”',
                self._ppp_phrase, self._pp_phrase, self._p_phrase)
        self._ppp_phrase = ''
        self._pp_phrase = ''
        self._p_phrase = ''
        self._new_sentence = False
```

There is a get_context() which tries to get the context again using surrounding text. get_context() currently doesn’t attempt to update self._new_sentence. I didn’t try that yet because surrounding text is so unreliable.

So currently it works only if you commit something which ends with a sentence end character and then do not move the cursor and continue typing the next word.

This should be improved of course but to really improve this, first surrounding text must work.

**Comment by psads-git** [comment]
That is fine, Mike, as one can write one's answers in `gedit` and paste them here.

**Comment by mike-fabian** [comment]
But in gedit the problem is the same. It works only if you don’t move the cursor.

**Comment by psads-git** [comment]
Thanks, Mike, but I had never noticed that!

**Comment by mike-fabian** [comment]
I just noticed that reopening preëdits works in firefox entry fields (for example here in github comments). In gedit it stopped working after recent ibus updates.

**Comment by psads-git** [comment]
> I just noticed that reopening preëdits works in firefox entry fields (for example here in github comments). In gedit it stopped working after recent ibus updates.

I can confirm that, Mike!

**Comment by mike-fabian** [comment]
I think the case insensitive match is an improvement, after testing this for a while I like it better than the old behaviour.

Maybe I could make an official release now...

**Comment by psads-git** [comment]
I also find the case-insensitive search an improvement, Mike.

For now, I have only one suggestion to offer. I use inline completion, but when all suggestions come from dictionaries, no inline completion is done. What do you think about this, Mike?

**Comment by mike-fabian** [comment]
I think this has nothing to do with whether the completions come from dictionaries or not.

Do you have a specific example where you think it doesn’t work as you think it should?

**Comment by psads-git** [comment]
For instance, see the screenshot below. Why is not `longe` offered as an inline completion?

![inline](https://user-images.githubusercontent.com/75945439/141987774-9e280315-a593-4891-bc40-76b555411e09.png)

**Comment by mike-fabian** [comment]
First of all `longe` is **not** a dictionary completion, it is from the user database. You see that because of the black colour and the star ⭐.  `longe` might **also** be in the dictionary, but apparently you typed it before already so it is in the user database.

Inline completions are only shown when the first candidate is an **exact** continuation of what you typed:

```
        if (not first_candidate.startswith(typed_string)
            or first_candidate == typed_string):
            # The first candidate is not a direct completion of the
            # typed string. Trying to show that inline gets very
            # confusing.  Don’t do that, show standard lookup table:
            self.update_lookup_table(self.get_lookup_table(), True)
            self._update_preedit()
            return
``` 

And in that case it is **not** an exact continuation of what you typed, you typed an uppercase L and the completion has a lower case l. 

(Before the case insensitive match, this wouldn't even have matched, the first match would have been `Lon` then but this would not have been shown as an inline completion either because there would have been nothing to complete, the preedit would have been exactly equal to the first candidate already in that case. 

I found it far to confusing when an inline completion changes anything the user has typed already. If such a change is there, better show the lookup table and have a  closer look.

**Comment by mike-fabian** [comment]
https://user-images.githubusercontent.com/2330175/141991259-cb4184d9-e803-4edb-9aa0-b3de3268c7a4.mp4

**Comment by psads-git** [comment]
OK, Mike. After having thought a bit about that matter, I have not reached any improvement.

I think you can go ahead with the new official release!

**Comment by mike-fabian** [comment]
Theoretically I could  change inline completion to allow changes in case or accents. If there is a bigger change than that by doing the inline completion, i.e. something like adding or removing letters or transposing letters, it is a serious spellchecking issue and changing this automatically by inline completion would be surprising. I am very upset when typing on the phone using SwiftKey and it replaces what I typed with something else. I have tried to disable this kind of “auto correction”.

But **maybe** one could make a case that changes from upper-case to lower case or adding or removing accents when doing the inline completion is “mostly harmless” and could be done with too many surprises. I guess it would still annoy me more than helping me but maybe I’ll try that sometime in the future. 

Currently  `egalit`, `Egalit`, `Ègalit` will not complete inline to `ègalite`. `ègalitè` will be among the candidates in the lookup table though. 

Wouldn’t it be weird if you type `Ègalit` and suddenly get the black part of the preedit change to `ègalit` followed by a gray `è`? This might destroy a capital letter one has typed on purpose because one is starting a new sentence for example and it would also interfere with auto-capitalizatioņ, sometimes lowercasing something to offer an inline completion when it was capitalized on pupose after a `.`. One could try to avoid that by doing a case changing inline completion only if `self._current_case_mode == 'orig'`, i.e. when the case has not been changed from what the user typed by auto-capitalization. 

When I first implemented inline completion and did not yet have the

```
        if (not first_candidate.startswith(typed_string)
            or first_candidate == typed_string):
            # The first candidate is not a direct completion of the
            # typed string. Trying to show that inline gets very
            # confusing.  Don’t do that, show standard lookup table:
            self.update_lookup_table(self.get_lookup_table(), True)
            self._update_preedit()
            return
```
it was extremely confusing and annoying to get the preedit changed to the first candidate **always** even if that changed the spelling a lot. 

So I had to add the above limitation to make it usable. **Maybe** the current limitation is too strict and there are a few more circumstances where it might be useful to allow inline completion. 

Would you really want that your input `Lon` gets completed to `lon`  + `ge`? 

Somehow I doubt this would be useful, but maybe I have to try it and see how it feels.

**Comment by psads-git** [comment]
Mostly, I agree with you Mike, but if the user writes 

`Lon`

why are not all suggestions capitalized?

**Comment by mike-fabian** [comment]
> Mostly, I agree with you Mike, but if the user writes
> 
> `Lon`
> 
> why are not all suggestions capitalized?

Adapting the case of the suggestions to the case the user typed may result in suggesting the wrong case. 

This is also a difficult problem.

When I type `portuguese` in English, it is probably wrong and should be capitalized to `Portuguese`.  In English this is written in upper case no matter whether `Portuguese` is used as a noun or as an adjective. 

In German however it depends on whether it is used as an adjective or as a noun:

“Ich spreche kein Portugiesisch.” (noun)
“Ich spreche nicht portugiesisch.” (adjective)

So when typing German, adapting the suggestion to the case the user typed will probably not make it worse, both could be correct. 

But in English, when typing `portuguese` adapting the suggestion `Portuguese` to he lower case the user typed  makes it worse. 

So whether suggestions should be converted automatically to the case typed by the user is unfortunately also a difficult question. I am not sure if they should. Always? Or only in some cases? If only in some cases, when exactly?

**Comment by psads-git** [comment]
In my opinion, Mike, if the user starts typing the word by a capital letter, the suggestions should _always_ be capitalized words. However, if the user starts typing in lowercase, then the suggestions should _mix_ (if appropriate) capitalized words with non-capitalized ones.

To type a capital letter is more costly than to type a non-capitalized one. Consequently, when an user chooses to start typing with an uppercase letter, then the user is meaning that wants only capitalized words.

**Comment by mike-fabian** [comment]
> In my opinion, Mike, if the user starts typing the word by a capital letter, the suggestions should _always_ be capitalized words. However, if the user starts typing in lowercase, then the suggestions should _mix_ (if appropriate) capitalized words with non-capitalized ones.
> 
> To type a capital letter is more costly than to type a non-capitalized one. Consequently, when an user chooses to start typing with an uppercase letter, then the user is meaning that wants _only_ capitalized words.
> 
> Does this make sense, Mike?

Maybe, I’ll think about it.

**Comment by psads-git** [comment]
OK, Mike. Thanks!

**Comment by mike-fabian** [comment]
I opened the new issue https://github.com/mike-fabian/ibus-typing-booster/issues/253 for this so that this idea does not get lost.

**Comment by psads-git** [comment]
Thanks, Mike. Have a good day!

**Comment by mike-fabian** [comment]
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/ has  2.15.1 builds which show only capitalized candidates if the user starts a word with a capital letter.

**Comment by psads-git** [comment]
Thanks, Mike! It seems to be working fine.

**Comment by mike-fabian** [comment]
> Thanks, Mike! It seems to be working fine.

Do you think it is an improvement? I tend to think it is an improvement but I am not sure yet...

**Comment by psads-git** [comment]
Yes, Mike, to me it is a big improvement. It is hard to understand why you are not yet sure about that. However, if you are not very convinced, you may want to ponder to make that optional.

**Comment by mike-fabian** [comment]
Ah, no, I don't want to make that optional, I wouldn't even know how to name that option so that it is understandable. 

I just didn't test enough yet. 

I thought there might be cases where one wants lower case even though one accidentally started typing with upper case. But that is probably rare and then one can still go to lower case by pressing the right Shift key. 

So I think I leave that feature. 

I might change the way it is implemented though, the current implementation makes it slightly slower, I have a different idea how it could be done.

**Comment by psads-git** [comment]
Thanks, Mike. I am afraid that my suggestions may be putting a heavy load on your shoulders. That is not my intention -- I just want to make ibus-typing-booster even greater, as  it simplifies a lot my life. However, if you do not feel like accepting or implementing my suggestions, you please feel free to follow your will and not my ideas and suggestions, as I understand that working on ibus-typing-booster is not the way your salary is earned. Moreover, you are the only one with full rights to decide what is best to Ibus-typing-booster.

**Comment by mike-fabian** [comment]
I think I might leave the current implementation of capitalizing the candidates when the user types a capital letter as it is.
It has a slight performance penalty because it does this:


https://github.com/mike-fabian/ibus-typing-booster/blob/release-candidate-2.15.1/engine/tabsqlitedb.py#L379

```
    def best_candidates(
            self,
            phrase_frequencies: Dict[str, int],
            title=False) -> List[Tuple[str, int]]:
        '''Sorts the phrase_frequencies dictionary and returns the best
        candidates.

        Should *not* change the phrase_frequencies dictionary!
        '''
        if title:
            phrase_frequencies_title: Dict[str, int] = {}
            for phrase in phrase_frequencies:
                phrase_title = phrase[:1].title() + phrase[1:]
                if phrase_title in phrase_frequencies_title:
                    phrase_frequencies_title[phrase_title] += (
                        phrase_frequencies_title[phrase_title])
                else:
                    phrase_frequencies_title[phrase_title] = (
                        phrase_frequencies[phrase])
            return sorted(phrase_frequencies_title.items(),
                          key=lambda x: (
                              -1*x[1],   # user_freq descending
                              len(x[0]), # len(phrase) ascending
                              x[0]       # phrase alphabetical
                          ))[:20]
        return sorted(phrase_frequencies.items(),
                      key=lambda x: (
                          -1*x[1],   # user_freq descending
                          len(x[0]), # len(phrase) ascending
                          x[0]       # phrase alphabetical
                      ))[:20]
```

I.e. when the function `best_candidates()` is called with `title=True`, then all phrases in the `phrase_frequencies` dictionary which are identical except for the capitalization of the first letter are merged into one and their user frequencies added.

For example, if phrase_frequencies contains

```
{'Ratatouille': 5, 'ratatouille' 3, 'ratatouiller': 1}
```

then the new phrase_frequencies_title will be:

```
{'Ratatouille': 8, 'Ratatouiller': 1}
```

and then this is sorted as usuall and the first 20 candidates returned.

Going through all then entries in the original dictionary, and merging into a new one is some extra work, especially if the original number of phrases before sorting and cutting out the first 20 is long, might be hundreds of entries long. But I don’t really notice any slowness while typing, so it is probably OK.

The other idea I had is this:

There is already the feature that you can switch between three case modes using the left and the right Shift key, forwards with the left one, backwards with the right one.

`'capitalize', 'upper', 'lower'`

and if you don’t like any of these you can go back to the `'orig'` case mode (original, as typed  by the user) by pressing Escape. This is somewhat faster as the above way to change the case in the candidate list as it only works on the 20 candidates which are already in the candidate  list and it just changes the case of them, if that creates identical entries, they are  not merged. For example if the candidate list contains:

```
1 Ratatouille
2 ratatouille
3 ratatouiller
```

and you press Shift you get:

```
1 Ratatouille
2 Ratatouille
3 Ratatouiller
``

so the first two entries are not identical. This is good because before pressing Shift you might have seen candidate 3 and thought “I want that but in upper case”, so you press `Shift` and then `3`. But if entries 1 and 2 were merged, the numbers would change and a candidate 3 would not exist anymore.

So switching through the case modes leaves the candidate list unchanged except for the change in case.

My other idea was, that when the user types a word starting with a capital letter to switch to the case mode `'capitalize'` automatically. That would be very fast but would often create duplicates in the candidate list shown.

My current implementation for the case  when the user typed a word starting with a capital letter gives you a different candidate list without duplicates. Therefore it might have a different length and different order because it gives higher weight to a word if it was already twice in the list with different capitalization.

I think removing the duplicates when the user starts typing with a capital letter is probably better.

**Comment by psads-git** [comment]
Now, I understand your reluctance regarding this new feature, Mike! But I was not thinking in anything as complicated as you describe!

Let me explain what was my idea. Nothing would need to change but the following:

- When the list of suggestions is ready to offer to the user (using the exact same mechanism as before, with _no_ modification at all), ibus-typing-booster should give it to the user as a list of capitalized words and removing the duplicates. 

This should not impose any degradation on speed, since it only implies 

- capitalize 9 words;

- and using a function `unique` to remove the duplicates.

**Comment by mike-fabian** [comment]
Removing the duplicates is sligthly more complicated then using a `unique` function on the words only because the candidates are pairs of words and integer numbers, higher number means higher priority, it should show up higher in the lookup table. 

When merging `{'Ratatouille': 5, 'ratatouille' 3}` into one entry, one needs to look at the numbers 5 and 3 as well. Either add them and get 8 (which is what I did and which increases priority of words which happened to be twice in the original list) or take the maximum. 

The final list presented to the user can have more than 9 entries because the candidate list can be paged down.
I cut it down to 20. If emojis are matched as well the final list can get longer than 20 again because the emoji are added. 

(You see that when you type `cat` you get 20, when typing `cat_`  you get 40 candidates)

Currently I remove the duplicates before cutting down to 20 candidates and resort because adding the frequencies of duplicates together requires a sort to get the entries in the right order of the frequencies.  But before cutting down to 20 the list may be long. 

I could make it slightly faster by first cutting down to 20 and sorting as usual and  **after** that remove the duplicates without changing the sort order (The first one of a duplicate is the one with the higher frequency if the list is already sorted). This will probably be slightly faster because one never goes through more then 20 candidates that way.

**Comment by psads-git** [comment]
> I could make it slightly faster by first cutting down to 20 and sorting as usual and **after** that remove the duplicates without changing the sort order (The first one of a duplicate is the one with the higher frequency if the list is already sorted). This will probably be slightly faster because one never goes through more then 20 candidates that way.

Right, Mike!

**Comment by mike-fabian** [comment]
I might try that, it could be that the difference in speed is not significant it seems, that current more “exact” implementation seems fast as well.

**Comment by psads-git** [comment]
Up to now, Mike, I have not noticed any slowdown. So, the exact approach seems to be appropriate.

**Comment by mike-fabian** [comment]
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/ has 2.15.2 builds now.

These builds have the sligthly faster implemtation of title casing the candidates if the user input starts with an upper case letter. 

The difference  is very small as I expected it is barely measurable.

But the new code is cleaner:

```
    def best_candidates(
            self,
            phrase_frequencies: Dict[str, int],
            title=False) -> List[Tuple[str, int]]:
        '''Sorts the phrase_frequencies dictionary and returns the best
        candidates.

        Should *not* change the phrase_frequencies dictionary!
        '''
        candidates = sorted(phrase_frequencies.items(),
                            key=lambda x: (
                                -1*x[1],   # user_freq descending
                                len(x[0]), # len(phrase) ascending
                                x[0]       # phrase alphabetical
                            ))[:20]
        if not title:
            return candidates
        candidates_title = []
        phrases_title = set()
        for candidate in candidates:
            phrase = candidate[0]
            phrase_title = phrase[:1].title() + phrase[1:]
            if phrase_title in phrases_title:
                continue
            candidates_title.append((phrase_title, candidate[1]))
            phrases_title.add(phrase_title)
        return candidates_title
```

Unless this makes anything worse, I’ll keep that.

**Comment by psads-git** [comment]
Thanks, Mike. Apparently, everything seems to be working fine.

Maybe in the future ibus-typing-booster will be able to work without recording uppercase words in the database and even faster. Notwithstanding, ibus-typing-booster is already fast and great!

**Comment by mike-fabian** [comment]
> Maybe in the future ibus-typing-booster will be able to work without recording uppercase words in the database and even faster. Notwithstanding, ibus-typing-booster is already fast and great!

I am not sure what you mean by “without reording uppercase words”.

Currently entries in the database look like this:

```
55897|mo|Mondays|hates|Charlie|2|1637249680.12463
```

The fields are

```
id input_phrase phrase p_phrase pp_prase  user_freq timestamp
```

The `input_phrase`, i.e. `mo` is what the user typed, it is already always recorded in lower case. That is how I implemented the case insensitive matching, input_phrase is always recorded in lower case now and new user input is also converted to lowercase before attempting to match it against `input_phrase` in that database.

`phrase` i.e. `Monday` is the candidate the user committed. Of course that has to be recorded with the case which was actually used, in this case upper case. I guess you don’t mean you want to change that.

`p_phrase` is the previous word and `pp_phrase` the word before that. I.e. the user typed `Charlie hates mo` and then selected `Mondays`.

This context of the last two words is currently recorded with the case which was actually used in the context.

That means if you type `Charlie hates mo` again, `Monday` will be suggested with a very high priority because the context matches exactly. Most likely it will be the first candidate.

However if you type `charlie hates mo`, `Monday` will still be in the suggestions because the `mo` matches and the `hates` match, but `charlie` does not match `Charlie` so the priority number calculated for that match will be somewhat smaller.

Now the question is how picky one should be about differences in case in the context. Should `Charlie hates` and `charlie hates` be considered to be the identical context or not?

Is there any need to make these to context distinct and possibly suggest different things in these two contexts?

Or should contexts which differ only in case be considered identical and produce exactly the same suggestions?

I am not sure what is better, it is possible that ignoring the case in the context is better, very hard to tell.

Punctuation for example is already ignored in contexts. Whether one types `(Charlie hates mo` or `Charlie hates mo`, the context is `Charlie hates` in both cases, the `(` is ignored. That makes it impossible to suggest different things in these two cases but probably suggesting something different just because one started with the `(` would not be helpful.

It might be that suggesting different things for contexts which differ only in case is not helpful either.

**If** suggesting different things when the the context differs only by case is not helpful **then** not recording the case of the context in the database would use up only one row instead of two in the database for `Charlie hates mo` and `charlie hates mo` with the result `Mondays`. As we limit the database to a fixed maximum number of 50000 rows, this would make room for a different entry, which might be more useful than storing **almost** identical entries which differ only in case.

I can try that, but I wonder how to test whether it is really better or decreases the quality of the suggestions.

**Comment by mike-fabian** [comment]
By the way, if we think about ignoring case in the context, we could also think about ignoring accents in the context. 

Both ignoring case and ignoring accents in the context makes the context less precise and therefore the suggestion less precise but it is unclear how much precision is helpful there. Too much precision can have the result that there are too many rows in the database which are never typed **exactly** like that again but something similar which differs only by case or accents is typed again. So a somewhat more “fuzzy” context could be better.

But it is really not easy to say when making such a change whether it improves things or makes things worse. 

I wonder if I should think about a way to measure improvement in a more precise way than just typing for a few hours and then deciding “It seems better to me ...”.

**Comment by psads-git** [comment]
Thanks, Mike. I think you have interpreted correctly my idea. I conjecture that, in this way, we can save database space without sacrificing prediction accuracy (or only very, very residually).

As a side comment, I have just checked GBoard (the smartphone Google keyboard): When one starts writing, some suggestions are shown and the same exact suggestions are turned capitalized if one starts writing with an uppercase letter!

**Comment by mike-fabian** [comment]
> Thanks, Mike. I think you have interpreted correctly my idea. I conjecture that, in this way, we can save database space without sacrificing prediction accuracy (or only very, very residually).


I can try that soon and make a test build and then you can try to test whether it is helpful or not.

Should I start with recording the context in lower case and still keep the accents or do both at the same time?

> As a side comment, I have just checked GBoard (the smartphone Google keyboard): When one starts writing, some suggestions are shown and the same exact suggestions are turned capitalized if one starts writing with an uppercase letter!

SwiftKey on Android also does this, so this is probably a good thing, we should keep that (and not make it optional either).

**Comment by psads-git** [comment]
I would not start by both at the same time, Mike. And I would start by the accents.

Maybe one could test the viability of the idea without implementing it in ibus-typing-booster: Doing some SQL queries _directly_ on the database would suffice.

**Comment by mike-fabian** [comment]
> I would not start by both at the same time, Mike. And I would start by the accents.
> 
> Maybe one could test the viability of the idea without implementing it in ibus-typing-booster: Doing some SQL queries _directly_ on the database would _perhaps_ suffice.

Implementing is not so difficult, I can probably do that in a few hours.

I wonder more how to evaluate whether it is helpful or not.

I just got this idea of a way of testing:

Letting ibus-typing-booster read whole books to train the input turned out to be less useful than I thought it would because some random book is not specific to the stuff a certain user usually types.

But of course I’ll keep that feature because there maybe some text files a user has which can be useful.

But letting ibus-typing-booster read some book might be a useful test case:

- Start with an empty database in memory
- Let ibus-typing-booster read some book into that database
- Start a simulated typing of the whole book letter by letter.
  I.e. type the first letter of the first word, see if the first word is suggested as the first candidate, if not type the next letter.
  As soon as the correct word is the first candidate, add the number of letters saved typing to some counter and continue with the next word of the book.
  “Type” the whole book like this and count how many letters were saved overall and calculate a “% of letters saved”.

Now change something in the implementation like storing the context case insensitive and repeat the above test and see whether the “% of letters saved” changes.

That might be a good test to see whether a change in implementation achieves an improvement or not.

**Comment by psads-git** [comment]
Go ahead, Mike! It seems to be a very ingenious testing idea!

**Comment by mike-fabian** [comment]
I implemented that test case now and checked what the difference is before and after this commit:

commit e252672cf50d3a778e6e243354702909065c9dd5
Author: Mike FABIAN <mfabian@redhat.com>
Date:   Wed Nov 17 12:47:30 2021 +0100

    Title case all candidates if input_phrase is in title case
    
    (Resolves: https://github.com/mike-fabian/ibus-typing-booster/issues/253)

I used the Project Gutenberg version of “The Picture of Dorian Gray” as a test case.

That book has 9195 lines and 80687 words:

```
$ wc the_picture_of_dorian_gray.txt 
  9195  80687 451987 the_picture_of_dorian_gray.txt
```

After reading that into an empty in memory database, that database has 70594 rows.

Testing with the behaviour **before** the above commit (i.e. even when a word starting with a capital letter is typed, there may be lower case candidates) gave:

```
total_length_typed=222279
total_length_committed=341048
total_length_saved=-118769
total_percent_saved=-34.82471675541273
```

I.e. 118769 characters were saved in typing which is 34.8%.

Testing **after** the above commit (When a word starting with a capital letter is typed, all candidates start with upper case) gave:

```
total_length_typed=216400
total_length_committed=341048
total_length_saved=-124648
total_percent_saved=-36.54852102929793
```

I.e. 124648 characters were saved in typing which is 36.5%.

So this change was helpful, although less than I would have guessed from my feelings after manual testing.

**Comment by mike-fabian** [comment]
If I tested correctly, storing the context only in lower case and matching the context only in lower case does not change the savings at all, I get **exactly** the same:

```
total_length_typed=216400
total_length_committed=341048
total_length_saved=-124648
total_percent_saved=-36.54852102929793
```

but it reduces the number of database rows stored from 70594 to 69743 (i.e. by 1.2%)

So it is a an improvement as it achieves exactly the same prediction quality with a somewhat smaller database.

But a very small improvement.

I’ll test with accent insensitive context next but I need to use a different book, “The Picture of Dorian Gray” has no accents at all.

**Comment by psads-git** [comment]
That is good news, Mike! I expect that the savings in the database size will be larger than in the case you have just studied.

**Comment by mike-fabian** [comment]
Actually I expect the savings by doing it without the accents to be even smaller than doing it case insensitive.

Let’s see ...

**Comment by mike-fabian** [comment]
I made some tests with “Notre-Dame de Paris by Victor Hugo”:

https://www.gutenberg.org/ebooks/2610

Results case **and** accent insensitive context:

```
Database rows: 156245
total_length_typed=619150
total_length_committed=849995
total_length_saved=-230845
total_percent_saved=-27.158395049382644

Results case insensitive context (but accent sensitive):

Database rows: 156301
total_length_typed=619150
total_length_committed=849995
total_length_saved=-230845
total_percent_saved=-27.158395049382644

Results case sensitive and accent sensitive context:
Database rows: 157587
total_length_typed=619150
total_length_committed=849995
total_length_saved=-230845
total_percent_saved=-27.158395049382644

Results before:
commit e252672cf50d3a778e6e243354702909065c9dd5
Author: Mike FABIAN <mfabian@redhat.com>
Date:   Wed Nov 17 12:47:30 2021 +0100

    Title case all candidates if input_phrase is in title case
    
    (Resolves: https://github.com/mike-fabian/ibus-typing-booster/issues/253)

total_length_typed=629175
total_length_committed=849995
total_length_saved=-220820
total_percent_saved=-25.978976346919687
```

I.e. just as with my tests with the English book “The Picture of Dorian Gray”, doing the “Title case all candidates if input_phrase is in title case” is definitely helpful although the difference is again not very big.

It only saves a small number of database rows.

Going to case insensitive context saves 1286 database rows (0.82% of the rows).

Doing accent insensitive context as well saves another 56 database rows (0.035% of the rows).

Doing the case sensitive and accent insensitive context does not change prediction accuracy at all in this test, the number of chracters saved is **exactly** the same.

When I think about it, this seems to make sense to me. It does not surprise me much that the context which is identical except for accents is very rare. Context which is identical except for the case of one of the context words will occur regularly depending whether a sequence of 3 words occurs at the beginning of a sentence or in the middle of the sentence. But I guessed already that context which is identical except for accent differences is something very unusual.

By the way, the 27.2% percent saved for “Notre-Dame de Paris by Victor Hugo”  was less than the 36.5% saved for “The picture of Dorian Gray”.

I am not sure why that is, maybe it Typing Booster works less well for French than for English? It could also be because of the size of the book, “Notre-Dame de Paris by Victor Hugo” is more than twice as long:

```
$ wc /home/mfabian/tmp/the_picture_of_dorian_gray.txt 
  9195  80687 451987 /home/mfabian/tmp/the_picture_of_dorian_gray.txt
$ wc /home/mfabian/tmp/victor_hugo_notre_dame_de_paris.txt
  21676  175408 1117074 /home/mfabian/tmp/victor_hugo_notre_dame_de_paris.txt
```

and the percentage saved tends to be bigger for shorter texts according to the tests I did yesterday.

Two shorter texts I tested yesterday:

```
$ wc the_road_not_taken.txt chant_d_automne.txt 
  27  151  770 the_road_not_taken.txt
  39  226 1396 chant_d_automne.txt
```

(“The Road Not Taken” is the poem by Robert Frost, “Chant d’automne” is the poem by Charles Baudelaird).

Savings were -51.3% for “The Road Not Taken” and -37.6% for “Chant d’automne”.

**Comment by mike-fabian** [comment]
So what do I do now after these tests?

I think I’ll do both the case insensitive **and** accent insensitive context.

The improvement doing this is small, it does not seem to change the prediction accuracy at all but it saves a (very small) amount of database rows. 

As the room in the database is limited (currently we cut it down to 50000 rows on each reastart of Typing Booster), saving a few rows without changing the prediction accuracy at all makes some room for other additional rows which might actually improve prediction accuracy.

So doing this seems to be an improvement, but a very small one.

**Comment by psads-git** [comment]
Excellent news, Mike! 

The following is perhaps a too crazy idea (I still have to think much deeper about it): Ignoring the order of the context words may have a very neglectable impact on prediction accuracy, while saving a lot of database rows.

**Comment by mike-fabian** [comment]
> Excellent news, Mike!
> 
> The following is perhaps a too crazy idea (I still have to think much deeper about it): Ignoring the order of the context words may have _only_ a very neglectable impact on prediction accuracy, while saving a lot of database rows.

Here I would guess that this will significantly worsen prediction accuracy.

That would destroy the whole trigram idea. Recently you qoted something which had some number how much prediction accuracy increases  if you use longer context than just 2 words, i.e. four-grams or five-grams.

According to what I have read so far, prediction accuraccy increases the longer you make the context used, but the speed goes down fast, the calculation time does increase extremely fast and the accuracy gains are minor. 

trigrams usually seem to be still worth it though. 

Your idea is almost like using bigrams with the additional twist of adding fake bigrams where the context skips one word.

I really doubt that this would help anything, most likely this makes it worse.

**Comment by psads-git** [comment]
As I remarked in my previous message, Mike, I have not yet thought thoroughly about the idea. Moreover, this is only a speculative exercise, since ibus-typing-booster is already fast -- it does not need significant improvements. However, my idea is to try to understand whether an ordered 3-gram model is worse than an _unordered_ 4- or 5-gram one.

**Comment by mike-fabian** [comment]
At least I have some code now which helps me testing such ideas. So we can get actual results now when trying out ideas.

Is there any literature about “unordered n-grams”?

**Comment by psads-git** [comment]
Yeah, your system to test new ideas is indeed a great improvement, Mike!

Yes, there little literature regarding unordered n-grams, but I have found the following technical report on

[Natural Language Generation](https://dspace5.zcu.cz/bitstream/11025/43167/1/Rigo_Sido_2020_05.pdf)

Unordered n-grams are mentioned on p. 7. However, as far as I understand, this technical report is not on word prediction.

**Comment by mike-fabian** [comment]
A colleague found this:

https://towardsdatascience.com/next-word-prediction-with-nlp-and-deep-learning-48b9fe0a17bf

but reading this I have the impression that they do **less** than I already do and that in a more complicated way.

They even write in a note:

> Note: There are certain cases where the program might not return the expected result. This is obvious because each word is being considered only once. This will cause certain issues for particular sentences and you will not receive the desired output. To improve the accuracy of the model you can consider trying out bi-grams or tri-grams. We have only used uni-grams in this approach. Also, a few more additional steps can be done in the pre-processing steps. Overall, there is a lot of scope for improvement.

**Comment by mike-fabian** [comment]
A found a side effect of making the user input into the database case insensitive and opened a new issue for this:

https://github.com/mike-fabian/ibus-typing-booster/issues/255

**Comment by mike-fabian** [comment]
New issue for the case and accent insensitivity in the context:

https://github.com/mike-fabian/ibus-typing-booster/issues/256

**Comment by mike-fabian** [comment]
I did another interesting test:

- Let Typing Booster read the book `victor_hugo_notre_dame_de_paris.txt` into an empty, in-memory database.

After doing that, the database has 156245 rows. We know already from the tests above that a simulated typing of that book will now save about -27% of the characters.

But that database is huge and will be cut down to the 50000 “best” entries on the next restart of Typing Booster. So how much will that degrade the prediction quality? So I do these next  steps in the test:

- Call cleanup_database() on that in-memory database
- Now the database has only 50000 entries
- Now do the typing simulation

Result: Only -24% are saved instead of -27%

But that doesn’t seem bad to me, only a small loss of 3% prediction accuracy with a database of less then 1/3 the original size.

**Comment by mike-fabian** [comment]
Getting higher quality data into a smaller database is more userful than a huge database which low quality data.

The database trained by the huge book “Notre dame de Paris” is probably mostly useful only to retype exactly that book and of very limited use to a “normal” user.

A “normal” user is probably much better served by a smaller database with contents which fit better to his style of writing.

**Comment by psads-git** [comment]
> A colleague found this:
> 
> https://towardsdatascience.com/next-word-prediction-with-nlp-and-deep-learning-48b9fe0a17bf
> 
> but reading this I have the impression that they do **less** than I already do and that in a more complicated way.
> 
> They even write in a note:
> 
> > Note: There are certain cases where the program might not return the expected result. This is obvious because each word is being considered only once. This will cause certain issues for particular sentences and you will not receive the desired output. To improve the accuracy of the model you can consider trying out bi-grams or tri-grams. We have only used uni-grams in this approach. Also, a few more additional steps can be done in the pre-processing steps. Overall, there is a lot of scope for improvement.

I do not think, Mike, it would be a good idea to use deep learning methods on ibus-typing-booster, as such a methods would require periodic training.

**Comment by psads-git** [comment]
> I did another interesting test:
> 
> * Let Typing Booster read the book `victor_hugo_notre_dame_de_paris.txt` into an empty, in-memory database.
> 
> After doing that, the database has 156245 rows. We know already from the tests above that a simulated typing of that book will now save about -27% of the characters.
> 
> But that database is huge and will be cut down to the 50000 “best” entries on the next restart of Typing Booster. So how much will that degrade the prediction quality? So I do these next steps in the test:
> 
> * Call cleanup_database() on that in-memory database
> * Now the database has only 50000 entries
> * Now do the typing simulation
> 
> Result: Only -24% are saved instead of -27%
> 
> But that doesn’t seem bad to me, only a small loss of 3% prediction accuracy with a database of less then 1/3 the original size.

Yes, Mike, that is great that saving a lot of database size impacts so little on accuracy!

**Comment by psads-git** [comment]
> Getting higher quality data into a smaller database is more userful than a huge database which low quality data.
> 
> The database trained by the huge book “Notre dame de Paris” is probably mostly useful only to retype exactly that book and of very limited use to a “normal” user.
> 
> A “normal” user is probably much better served by a smaller database with contents which fit better to his style of writing.

I totally agree with you, Mike!

**Comment by mike-fabian** [comment]
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/ has 2.15.4 now with these changes:

- Make the context in the database case insensitive and accent insensitive (Resolves: https://github.com/mike-fabian/ibus-typing-booster/issues/256)
- Test cases for the sqlite database
- Allow cleanup_database() to cleanup an in memory database when doing unit tests
- Allow only lower case user shortcuts (Resolves: https://github.com/mike-fabian/ibus-typing-booster/issues/255)
- Add function to return number of rows in the database for debugging and testing
- Allow to read training data from file when database is empty, allow reading from .gz files

**Comment by psads-git** [comment]
Thanks, Mike. If I find any problem, I will let you know.

**Comment by mike-fabian** [comment]
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/ The 2.15.7 build contains an additional small tweak:

While reading training data from a file, the context in the database is converted to lower case and no accents.

So if you want to convert the context in the existing rows in your  database you can read trainig data from file.

Size of the file doesn’t matter, it can even be empty.

**Comment by psads-git** [comment]
Thanks, Mike. That is an useful tweak!

**Comment by mike-fabian** [comment]
What do you think about this?:

https://github.com/mike-fabian/ibus-typing-booster/issues/257

Do you have any opinions? If yes, please comment.

**Comment by psads-git** [comment]
I do not think that it is a high-needed feature, Mike, given the fact that it only needs to copy a file. But, who knows, it may be useful for non sophisticated users. 

If you choose to add such a feature, then maybe it should include not only the database, but all configurations. And maybe using a compressed file as the exported file.

I hope to have helped, Mike!

**Comment by mike-fabian** [comment]
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/ has 2.15.8 builds.

Almost everything works 30%-40% faster with this experimental version (but uses somewhat more memory, not sure how much)

**Comment by psads-git** [comment]
Thanks, Mike. With _uses somewhat more memory_, do mean RAM or disk?

**Comment by mike-fabian** [comment]
RAM.

I didn't measure how much more ram it uses and I am not sure how to measure that.  

The change which achieves this big speedup is actually only two lines:

```
diff --git a/engine/itb_util.py b/engine/itb_util.py
index ea55b7b7..9b4cfee6 100644
--- a/engine/itb_util.py
+++ b/engine/itb_util.py
@@ -33,6 +33,7 @@ from enum import Enum, Flag
 import sys
 import os
 import re
+import functools
 import collections
 import unicodedata
 import locale
@@ -2784,6 +2785,7 @@ TRANS_TABLE = {
     ord('Ŧ'): 'T',
 }
 
+@functools.cache
 def remove_accents(text: str, keep: str = '') -> str:
     # pylint: disable=line-too-long
     '''Removes accents from the text
```

I noticed that the function which removes accents froma a string is a major bottleneck.

I couldn’t find a way to make that function faster but I tried caching the results.
The easiest way to do that is to add that function decorator.

That means if this function is called twice with the same arguments, for example if you call something like this twice:


```
remove_accents('abcÅøßẞüxyz', keep='åÅØø'))
```

then the second call with return the result

```
'abcÅøssSSuxyz'
```

from the cache which is of course much faster.

As this `remove_accents()` function is used really a lot in Typing Booster, caching results from only that function already achieves that huge speedup of 30%-40%.

But as I didn’t limit the size of the cache, this means that **every** time this function is called with a different word during a typing booster session, that word gets added to the cache. And this function is really used a lot, i.e. the cache might get quite big.

I could use something like:

```
@functools.lru_cache(maxsize=100000)
```

to limit the maximum size of the cache. That would limit the cache to up to the most recent one hundred thousand calls. As each call typically has a word of input and a word of output, that would limit the maximum size of the cache to a few megabytes.

According to the documentation

https://docs.python.org/3/library/functools.html

adding such a limit makes it slightly slower though:

> Returns the same as lru_cache(maxsize=None), creating a thin wrapper around a dictionary lookup for the function arguments. Because it never needs to evict old values, this is smaller and faster than lru_cache() with a size limit.

I have not yet measured how much slower.

**Comment by psads-git** [comment]
Thanks, Mike, for the clarification.

Let us see how much RAM does the cache consumes. But there is an upper bound: The database size, which is _only_ about 3.5MB!

**Comment by mike-fabian** [comment]
No, the database doesn’t really have much do do with it. The user input and the context are already stored without accents in the database. So for reading from the database, accents don’t need to be removed.

But while the user types and candidates are suggested, each time the user types an additional letter and new candidates are suggested, the `remove_accents()` function is called 3 times, on the current user input and on the 2 words of context. As the two words of context don’t change while the user types the current word, only for the first letter typed the `remove_accents()` function might actually do some work to remove accents from the context, for the next letters typed `remove_accents()` gets the results of the words from the context without accents from the cache already. That makes it faster and I feel the difference is big enough that I notice it while typing.

Each time the user has finished typing a word a new row is added or an existing row is updated. That causes the `remove_accents` function to be called 3 times.

During startup of Typing Booster, dictionaries are read. In case of languages using accents like Portuguese, German, ... the accents are removed from all words read from the dictionary.
This are typically a few ten thousand words per dictionary, 43887 words in case of Portuguese:

```
$ wc /usr/share/myspell/pt_PT.dic 
  43887   87788 1459095 /usr/share/myspell/pt_PT.dic
$ wc /usr/share/myspell/de_DE.dic 
  75782   75852 1101641 /usr/share/myspell/de_DE.dic
$
```

If each word is on average 5 characters long, input and output would be 10 characters on average, plus some overhead when storing in the cache, so probably the cache needs about 1 MB for a typical dictionary. That seems very reasonable to me. And it is unlikely that it grows significantly after that as the number of words a user types until the session is restarted will be much smaller than that. And all the words the user types which were already in the dictionary, are already in the cache and don’t make the cache grow.

**Comment by psads-git** [comment]
Thanks, Mike, for your clarification!

My reasoning is as follows. The need to remove accents arises from the fact that the database works with non-accented words. Hence, by absurd, one could put the entire database with all accented words on memory, with no trouble removing accents -- so the cache size is bounded by the size of the database with all words accented! 

I have meanwhile had an idea that _may_ dispense with any cache. As the user types a word, letter by letter, there is no need to remove the accents of the ongoing word but only of the last character _if_ accented!

**Comment by mike-fabian** [comment]
I think 1 MB of cache for this accent removal is no problem.

**Comment by psads-git** [comment]
> The database has the words without accents to make it possible to match he user input without accents. So I don’t understand what you mean by "the database with all accented words". There are no accented words in the database except for the resulting candidates.

Yes, Mike, but if you worked with accented context in the database records, there would not be any need to remove accents. Note, though, Mike, that I am not advocating that you use accented context -- I am completely on the opposite side. This is just a mental exercise to show that, at most, caching is limited to the size of the database. 

> Converting the user input character by character is also not possible because sometimes accents get created by typing several characters, depending on the input methods used. For example if t-latn-post is used, an `ü` can be typed by typing `u` followed by `"`, both are ASCII. Only when the final word is produced by transliteration the `ü` appears.

Sorry, Mike, but I was ignorant of that detail.

> I think 1 MB of cache for this accent removal is no problem.

I agree, Mike!

**Comment by mike-fabian** [comment]
We even see in the build times at
https://copr.fedorainfracloud.org/coprs/mfabian/ibus-typing-booster/builds/

that it is faster. 2.15.8 and 2.15.9 did build in 4 minutes, builds for older versions took 5-6 minutes.

This is the time saved in doing the test cases during the build. In real use the savings should be even greater because the build times include also the setup of the build (installing required packages into the chroot etc. ...), the building and the packing of the results into the rpms.

**Comment by mike-fabian** [comment]
> Note, though, Mike, that I am not advocating that you use accented context -- I am completely on the opposite side.

Maybe we should revisit this issue:

https://github.com/mike-fabian/ibus-typing-booster/issues/231

and think whether any fine grained configurability of accent insensitiv matching is helpful at all.

What we discussed there is quite a complicated change, with a complicated user interface to set it up which most users might not even understand.

And according to the tests I did recently, it will probably accomplish very little in improving the predictions. Doing completely accent insensitive predictions (accent insensitive user input and context) did not make the percent of characters saved any worse when I tested with the French book “Notre Dame de Paris”. There may be some cases where typing the correct accents would reduce the number of candidates a bit and one would get the correct candidate a bit earlier, but that really doesn’t seem to be the case often. If that happened often, probably it would have made the percent of characters saved worse when testing with that French book. When doing this test with the French book, I had always perfect context, always the two previous words were remembered correctly in the database. In reality, this is not always the case. Sometimes because surrounding text didn’t work and maybe the fallback to remember the last two words didn’t help either because the cursor was moved or the focus was moved to a different window. In reality, it happens more often that the context is missing or even wrong than when doing such a test of reading from a book and then retyping that exact book. When the context is empty, typing the accents correctly of the current input might help a bit more compared to when two words of perfect context are available.

I am not 100% sure, but I need to look at this again carefully. Maybe it is not worth doing what we discussed in https://github.com/mike-fabian/ibus-typing-booster/issues/231

**Comment by psads-git** [comment]
Given the improvements and the knowledge meanwhile learned, I tend to agree with your arguments, Mike.
